### PR TITLE
Start Always Encrypted feature branch

### LINKED ISSUE #56: Always Encrypted - get it working
- What is the correct connection string for Always Encrypt?
- How to integrate with Key Vault?
- Any example to Insert or Update the data?

I have an issue with updating an encrypted field. 
For example:
My table `members` has an column `dob` which is encrypted

I can exec it in Azure studio, but cannot implement in code.
```
DECLARE @dob varchar(10) = '2022-06-01';
Update members SET dob  = @dob where id = '123';
```

I always get `The encryption scheme for the columns/variables is (encryption_type = 'PLAINTEXT') and the expression near line '2' expects it to be DETERMINISTIC, or PLAINTEXT.`

Thanks

### LINKED ISSUE #104: Always Encrypted - define connection string parameters
Include the full set of capabilities including enclaves and per-key-provider authentication.

### LINKED ISSUE #105: Always Encrypted - define interface for key provider packages
The connection string will allow the app to identify which key provider to use for a connection.
The app can import individual key provider packages. Each imported package will register itself with the core driver.
The interface will have methods to allow the provider to parse relevant parameters from the connection string and to provide keys for encryption/decryption as needed.

### LINKED ISSUE #107: Always Encrypted - implement column decryption
see if Microsoft already has a golang implementation of AEAD-AES-256-CBC-HMAC-SHA-256

### PR DESCRIPTION
**Pull request by shueybubbles**
Fixes #107 
Fixes #105 
Fixes #104 
Fixes #56 

Key design tenets:
1. Key provider packages will sit outside the core module for apps to import as needed.
2. Decrypted column encryption keys will be cached for 2 hours by default. Key provider implementations can override the behavior by returning a zero duration to tell the driver not to cache its keys or by providing a custom duration.
3. Key provider implementations are registered globally. An application can provide a custom provider instance on a per-connection basis by adding it a `Connector` instance. 
4. Key provider parameters are NOT available from the connection string. Each provider package must implement its own configuration management. For example - the globally registered AKV provider will use `AzureDefaultCredential` for authentication, which allows an app to control authentication through completely external means. Apps that want to use a `ServicePrincipalCredential` explicitly in code can attach an instance of the AKV provider to the `Connector` instead of relying on the globally registered instance. That technique will allow different credentials per connection.

Design issues to resolve in future PRs
1. How to implement inserts of encrypted data in an idiomatic way. The current app model relies on implicit type conversion from Go types to SQL types and the driver often relies on the server to coerce the input type to the column type. That type of conversion will fail for encrypted data, as the SQL type sent from the client has to be an exact match for the column type.
2. Proper analysis and management of the key cache. It could become a bottleneck depending on the type of mutex and the application workload. I am likely to implement 3 cache models based on each of the standard Go primitives - `sync.Mutex, sync.RWLock, sync.Map` so the application can choose which is appropriate for its workload. Such a model is likely to lead to apps being able to provide their own cache implementation.

### CODE REVIEW

**Comment by shueybubbles** [comment]
Notes for code reviewers:
- Start with `alwaysencrypted_windows_test.go` to see the E2E test. It will give you the overview of the functionality for encrypt and decrypt that this PR creates for application use.
- Bulk insert of encrypted values is not implemented yet; you'll see a TODO for that.
- The swisscom code has been copied to this repo with its license preserved. The only addition to that code here is implementation of `Encrypt`.
- I'd love for a Linux expert to contribute an E2E test for the `pfx` provider on a Linux build agent. See `internal/certs.go` for the test infrastructure pieces that create and delete key encryption certificates.
- Char/varchar parameter data is a bit of a mess in this driver. It just sends the raw UTF8 bytes in the parameter, with no code page data. That approach doesn't fly for AE. I've opened an issue #129

File: alwaysencrypted_windows_test.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+		sel.WriteString(fmt.Sprintf("col%d,", i))
+	}
+	_, _ = query.WriteString("unencryptedcolumn nvarchar(100)")
+	_, _ = query.WriteString(")")
+	insertArgs[len(encryptableColumns)] = "unencryptedvalue"
+	insert.WriteString(fmt.Sprintf("@p%d)", len(encryptableColumns)+1))
+	sel.WriteString(fmt.Sprintf("unencryptedcolumn from [%s]", tableName))
+	_, err = conn.Exec(query.String())
+	if err != nil {
+		t.Fatalf("Failed to create encrypted table %s", err.Error())
```
Comment (shueybubbles) [inline alwaysencrypted_windows_test.go:113]: >t.Fatalf("Failed to create encrypted table %s", err.Error()) [](http://example.com/codeflow?start=2&length=60)

I've not used stretchr yet in my code, but will use stretchr asserts going forward. They warm my TDD heart. The code copied from swisscom uses it.

File: internal/github.com/swisscom/mssql-always-encrypted/pkg/algorithms/aead_aes_256_cbc_hmac_sha256.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	authTag := ciphertext[idx : idx+a.keySizeBytes]
+	idx += a.keySizeBytes
+
+	iv := ciphertext[idx : idx+a.blockSizeBytes]
+	idx += len(iv)
+
+	realCiphertext := ciphertext[idx:]
+	ourAuthTag := a.prepareAuthTag(iv, realCiphertext)
+
+	if bytes.Compare(ourAuthTag, authTag) != 0 {
```
Comment (qmuntal) [inline internal/github.com/swisscom/mssql-always-encrypted/pkg/algorithms/aead_aes_256_cbc_hmac_sha256.go]: Using a non constant-time compare for validating the authentication tag can leak information about the key. Use [subtle.ConstantTimeCompare](https://pkg.go.dev/crypto/subtle#ConstantTimeCompare) instead. #Closed

**Review by qmuntal — CHANGES_REQUESTED** [review]

File: quoter.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+func (TSQLQuoter) ID(name string) string {
+	return "[" + strings.Replace(name, "]", "]]", -1) + "]"
+}
+
+// Value quotes database values such as string or []byte types as strings
+// that are suitable and safe to embed in SQL text. The returned value
+// of a string will include all surrounding quotes.
+//
+// If a value type is not supported it must panic.
+func (TSQLQuoter) Value(v interface{}) string {
```
Comment (stuartpa) [inline quoter.go:22]: I don't think this code used?

Comment (shueybubbles) [reply quoter.go:22]: The interface is defined by `sqlexp` and can be used by applications.

File: alwaysencrypted_windows_test.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+	s := fmt.Sprintf(createColumnMasterKey, certPath, certPath)
+	if _, err := conn.Exec(s); err != nil {
+		t.Fatalf("Unable to create CMK: %s", err.Error())
+	}
+	defer conn.Exec(fmt.Sprintf(dropColumnMasterKey, certPath))
+	r, _ := rand.Int(rand.Reader, big.NewInt(1000))
+	cekName := fmt.Sprintf("mssqlCek%d", r.Int64())
+	tableName := fmt.Sprintf("mssqlAe%d", r.Int64())
+	keyBytes := make([]byte, 32)
+	_, _ = rand.Read(keyBytes)
```
Comment (apoorvdeshmukh) [inline alwaysencrypted_windows_test.go:71]: Should we assert err to be nil for Read and Exec (incase dropping table fails)? #Closed

**Review by apoorvdeshmukh — APPROVED** [review]
Looks good to me.
The testcase serves the E2E behavior but can we also add an example in the examples directory?

File: alwaysencrypted_windows_test.go (hunk 4 above)
Comment (shueybubbles) [reply alwaysencrypted_windows_test.go:71]: there's not much the test can do if drop table fails.
If for some reason it does fail, the subsequent create table will fail with a duplicate name and the test will fail anyway

rand.Read won't fail

**Pull request merged**